File: apps/services/notifications/Notifications.Infrastructure/Services/UsageMeteringService.cs
using Microsoft.Extensions.Logging;
using Notifications.Application.Interfaces;
using Notifications.Domain;

namespace Notifications.Infrastructure.Services;

public class UsageMeteringService : IUsageMeteringService
{
    private readonly IUsageMeterEventRepository _repo;
    private readonly ITenantBillingPlanRepository _billingRepo;
    private readonly ITenantBillingRateRepository _rateRepo;
    private readonly ILogger<UsageMeteringService> _logger;

    public UsageMeteringService(IUsageMeterEventRepository repo, ITenantBillingPlanRepository billingRepo, ITenantBillingRateRepository rateRepo, ILogger<UsageMeteringService> logger)
    {
        _repo = repo;
        _billingRepo = billingRepo;
        _rateRepo = rateRepo;
        _logger = logger;
    }

    public async Task MeterAsync(MeterEventInput input)
    {
        try
        {
            var (isBillable, currency) = await EvaluateBillingAsync(input.TenantId, input.UsageUnit, input.Channel, input.ProviderOwnershipMode);
            await _repo.CreateSilentAsync(new UsageMeterEvent
            {
                TenantId = input.TenantId,
                NotificationId = input.NotificationId,
                NotificationAttemptId = input.NotificationAttemptId,
                Channel = input.Channel,
                Provider = input.Provider,
                ProviderOwnershipMode = input.ProviderOwnershipMode,
                ProviderConfigId = input.ProviderConfigId,
                UsageUnit = input.UsageUnit,
                Quantity = input.Quantity,
                IsBillable = isBillable,
                ProviderUnitCost = input.ProviderUnitCost,
                ProviderTotalCost = input.ProviderTotalCost,
                Currency = input.Currency ?? currency,
                MetadataJson = input.Metadata != null ? System.Text.Json.JsonSerializer.Serialize(input.Metadata) : null,
                OccurredAt = DateTime.UtcNow
            });
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "[Metering] Failed to record meter event: {UsageUnit}", input.UsageUnit);
        }
    }

    public async Task MeterBatchAsync(IEnumerable<MeterEventInput> inputs)
    {
        foreach (var input in inputs)
            await MeterAsync(input);
    }

    private async Task<(bool IsBillable, string? Currency)> EvaluateBillingAsync(Guid tenantId, string usageUnit, string? channel, string? providerOwnershipMode)
    {
        try
        {
            var plan = await _billingRepo.FindActivePlanAsync(tenantId);
            if (plan == null) return (false, null);

            var rate = await _rateRepo.FindRateAsync(plan.Id, usageUnit, channel, providerOwnershipMode);
            if (rate != null) return (rate.IsBillable, rate.Currency);

            var fallbackRate = await _rateRepo.FindRateAsync(plan.Id, usageUnit);
            if (fallbackRate != null) return (fallbackRate.IsBillable, fallbackRate.Currency);

            return (false, plan.Currency);
        }
        catch
        {
            return (false, null);
        }
    }
}

File: _archived/notifications-nodejs/src/services/usage-metering.service.ts
import { UsageMeterEventRepository, CreateUsageMeterEventInput } from "../repositories/usage-meter-event.repository";
import { evaluateBilling } from "./billing-evaluation.service";
import { logger } from "../shared/logger";
import { UsageUnit, NotificationChannel } from "../types";

const repo = new UsageMeterEventRepository();

export interface MeterEventInput {
  tenantId: string;
  usageUnit: UsageUnit;
  channel?: NotificationChannel | null;
  notificationId?: string | null;
  notificationAttemptId?: string | null;
  provider?: string | null;
  providerOwnershipMode?: string | null;
  providerConfigId?: string | null;
  quantity?: number;
  providerUnitCost?: number | null;
  providerTotalCost?: number | null;
  currency?: string | null;
  metadata?: Record<string, unknown>;
}

/**
 * Records a usage meter event with billing evaluation.
 * Safe to call from anywhere in the send flow — never throws.
 */
export async function meter(input: MeterEventInput): Promise<void> {
  try {
    const billing = await evaluateBilling({
      tenantId: input.tenantId,
      usageUnit: input.usageUnit,
      channel: input.channel,
      providerOwnershipMode: input.providerOwnershipMode,
    });

    const payload: CreateUsageMeterEventInput = {
      tenantId: input.tenantId,
      notificationId: input.notificationId ?? null,
      notificationAttemptId: input.notificationAttemptId ?? null,
      channel: input.channel ?? null,
      provider: input.provider ?? null,
      providerOwnershipMode: input.providerOwnershipMode ?? null,
      providerConfigId: input.providerConfigId ?? null,
      usageUnit: input.usageUnit,
      quantity: input.quantity ?? 1,
      isBillable: billing.isBillable,
      providerUnitCost: input.providerUnitCost ?? null,
      providerTotalCost: input.providerTotalCost ?? null,
      currency: input.currency ?? billing.currency ?? null,
      metadataJson: input.metadata ? JSON.stringify(input.metadata) : null,
      occurredAt: new Date(),
    };

    await repo.createSilent(payload);
  } catch (err) {
    // Metering must never crash the send flow
    logger.error("[Metering] Failed to record meter event", { usageUnit: input.usageUnit, error: String(err) });
  }
}

/**
 * Batch meter multiple events at once.
 */
export async function meterBatch(inputs: MeterEventInput[]): Promise<void> {
  await Promise.allSettled(inputs.map(meter));
}
